returns/primitives/tracing.py

```diff
@@ -1,13 +1,12 @@
 import types
 from contextlib import contextmanager
 from inspect import FrameInfo, stack
-from typing import List, Optional
+from typing import Iterator, List, Optional
 
 from returns.result import _Failure
 
 
-@contextmanager
-def collect_traces():
+def collect_traces() -> Iterator[None]:
```

Comment:
Will this work? 

```python
from contextlib import contextmanager
from typing import overload, Iterator, Callable, TypeVar, TYPE_CHECKING, ContextManager, Optional, Union

if not TYPE_CHECKING:
    reveal_type = print

F = TypeVar('F', bound=Callable)


@overload
def collect_traces(f: F) -> F:
    ...

@overload
def collect_traces() -> ContextManager[None]:
    ...

def collect_traces(f: Optional[F] = None) -> Union[F, ContextManager[None]]:
    @contextmanager
    def factory() -> Iterator[None]:
        try:
            yield
        finally:
            print('done')
    return f if f is not None else factory()

@collect_traces
def some(a: int) -> str:
    return str(a)

with collect_traces():
    ...

reveal_type(some(1))
```

mypy is happy about it:

```
» mypy ex.py
ex.py:34: note: Revealed type is 'builtins.str'
```

Reply:
I've updated the implementation using your suggestion:

```python
def collect_traces(
    function: Optional[Function] = None,
) -> Union[Function, ContextManager[None]]:
    def factory() -> Iterator[None]:
        unpatched_get_trace = getattr(_Failure, '_get_trace')
        substitute_get_trace = types.MethodType(_get_trace, _Failure)
        setattr(_Failure, '_get_trace', substitute_get_trace)
        try:
            yield
        finally:
            setattr(_Failure, '_get_trace', unpatched_get_trace)
    collect_traces_cm = contextmanager(factory)()
    return (
        collect_traces_cm(function) if function else collect_traces_cm
    )
```

Some explanations:
```python
collect_traces_cm = contextmanager(factory)()
```
To avoid mypy error I've decided to pass the factory function directly to `contextmanager` function instead of using it as a decorator

---

```python
return (
    collect_traces_cm(function) if function else collect_traces_cm
)
```
`contextmanager(factory)()` returns an instance of `_GeneratorContextManager` class that implements `__enter__`, `__exit__` and `__call__`. This means to use like a common context manager we just return it, but to use it like decorator we have to pass trough the callable instance!